packages/core-data/src/types/wp_template.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * Internal dependencies
+ */
+import { Context, PostStatus, RenderedText, OmitNevers } from './common';
+
+interface FullWpTemplate< C extends Context > {
+	/**
+	 * ID of template.
+	 */
+	id: string;
+	/**
+	 * Unique slug identifying the template.
+	 */
+	slug: string;
+	/**
+	 * Theme identifier for the template.
+	 */
+	theme: string;
+	/**
+	 * Type of template.
+	 */
+	type: string;
+	/**
+	 * Source of template
+	 */
+	source: string;
+	/**
+	 * Source of a customized template
+	 */
+	origin: string;
+	/**
+	 * Content of template.
+	 */
+	content:
+		| {
+				/**
+				 * Content for the template, as it exists in the database.
+				 */
+				raw?: string;
+				/**
+				 * Version of the content block format used by the template.
+				 */
+				block_version?: number;
+		  }
+		| string;
```

Comment:
what can we say about this one? this is the one you said should not be updated as `{ content: string }` but rather it needs to be updated as `{ content: { raw: string } }`??

Reply:
> this is the one you said should not be updated as { content: string } but rather it needs to be updated as { content: { raw: string ??

No, that one is `Widget`. This here is modeled after what the schema says:

```js
{
  "description": "Content of template.",
  "type": [
    "object",
    "string"
  ],
  // ...
}
```

But looking at the code, I think it's always outputted as an object:
 
https://github.com/WordPress/WordPress/blob/7a48bab20c00819d7c30732592e8ed9eafd7af51/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L586-L595

I believe the schema says it may be a `string` because it accepts both string and arrays as an input:

https://github.com/WordPress/WordPress/blob/7a48bab20c00819d7c30732592e8ed9eafd7af51/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L504-L510

I'll remove the string from here and model it like `Post`

Reply:
Hm the `rendered` field is missing so we can't the `Updatable` type we have wouldn't collapse that into a string. I added a new branded type to handle this case:

```ts
export type StringWhenUpdatable< T > = T & { __brand: 'UpdatableAsString' };
interface FullWpTemplatePart< C extends Context > {
	content: StringWhenUpdatable< {
		/**
		 * Content for the template, as it exists in the database.
		 */
		raw: ContextualField< string, 'view' | 'edit', C >;
		/**
		 * Version of the content block format used by the template.
		 */
		block_version: ContextualField< number, 'edit', C >;
	} >;
}
```

And adjusted the `Updatable` definition like this:

```ts
export type Updatable< T extends EntityRecord< 'edit' > > = {
	[ K in keyof T ]: T[ K ] extends
		| RenderedText< any >
		| StringWhenUpdatable< any >
		? string
		: T[ K ];
};
```

I was also noodling on different names like `CollapsibleField` or `RenderedField` but I don't find them as clear.

Reply:
I ended up using this last approach instead of `StringWhenUpdatable`.

Reply:
> One thing I don't like about that is that we now have a __brand property. I can live with that, especially if we document it, but maybe there's an easy way to get rid of it?

generally we won't be creating these extra fields but rather we'll be casting _as_ the branded type. for now maybe it's best to leave this awkwardness in so we can think more about how to handle the template bit. it would be good to avoid deeply-coupling the types of the entities and `Updatable` if possible.

```
/** This field should be set as a string when sending updates to
 *  the server but it never has a `rendered` property when reading.
 *
 * @TODO: Figure out how to mesh this with `RenderedText<C>`
 */
content: (RenderedText<C> & { block_version?: … }) | string;
```

Reply:
Sounds good, I removed the `| string` from the snippet though as `Updatable` takes care of that.